create an entity relationship diagrams of the schemas using erdantic

yves-renier opened the pull request:
Hello,

I wanted to have ERD diagram for my models so I implemented this.
I added an option model-erdantic-figure to switch it off (it is currently on by default, but feel free to change that).
I also added the documentation for the option.

mansenfranzen commented:
Hi @yves-renier,

thanks for you PR! You did a great job diving deeply into `sphinx` and `autodoc_pydantic` internals ;-). 

I really like the idea of adding the possibility to show ERDs for pydantic models. I'm happy to support this PR.

So far, there remain some ToDos that we should address first:

- [ ] `autodoc_pydantic` will need to add `erdantic` as an dependency in its `pyproject.toml`. Otherwise, the docs build process runs into an import error.

- `erdantic` depends on `pygraphviz` and `graphviz` which will complicate the installation process of `autodoc_pydantic`.
  - `pygraphviz` does not offer wheels and [requires a C/C++ compiler](https://pygraphviz.github.io/). 
  - `graphviz` is not a python but external system dependency which can't be installed via pip.
  - Using `conda` here would simplify the process of setting up the doc build environment but I would rather prefer to keep it plain poetry/pip.
 - [ ] Hence, I suggest that `erdantic` should be an optional dependency. Accordingly, the ERD feature should be turned off by default.
 - [ ] If ERD is enabled, we should add a check to test against appropriate dependencies and provide a meaningful error message to inform users.
    - [ ] This will require a short documentation on how to properly set up `autodoc_pydantic` with `erdantic`, `pygraphviz` and `graphviz`.
- [ ] Since this is a great feature, I would very much like to see a dedicated example in the [example section](https://autodoc-pydantic.readthedocs.io/en/stable/users/examples.html).
- [ ] Currently, there are no tests for the new feature. This should be added. I can assist here since this is rather complicated due to the nature of `sphinx` tests. 

Please feel free to comment on my thoughts here. I'm happy to discuss them.

yves-renier commented:
Hello, thanks for the support !
I just added `erdantic` as an optional dependency (`pip install autodoc_pydantic[erd]`) and turned off the ERD feature by default.
I will look to add an example, check and document the dependencies and try to setup the tests locally.

mansenfranzen commented:
Thanks for continuing with the PR! 

If you get stuck or need any help, please don't hesitate to reach out.

yves-renier commented:
About example, I already added one at `/docs/build/html/users/configuration.html#show-erdantic-figure`.
I will also enable the option for the [complete example](https://autodoc-pydantic.readthedocs.io/en/stable/users/examples.html#complete).

yves-renier commented:
I just added a simple test.
@mansenfranzen, I believe I finished implementing your TODO.
I also modified the github actions to install erdantic and graphviz.

mansenfranzen commented:
@yves-renier The PR looks great - well done!

I can confirm that tests pass and documentation is rendered properly. I have a few thoughts remaining that prevent me from merging, yet:

- I dislike `erdantic`'s watermark at the bottom of the diagram. It distracts. I would rather see it being removed by `erdantic` or by `autodoc_pydantic`. The latter option is of course less desirable. I opened an issue [here](https://github.com/drivendataorg/erdantic/issues/78).
- The resulting diagram can be improved if rendered as SVG instead of PNG. SVG is superior because it is a vector and additional information is shown upon hovering over attributes of pydantic models. `sphinx.ext.graphviz` already provides a configuration to switch to SVG, see [here](https://www.sphinx-doc.org/en/master/usage/extensions/graphviz.html#confval-graphviz_output_format). We could make this the default for `autodoc_pydantic`.
- While seeing the generated diagrams in action for the first time in the newly build docs, I think the diagrams can be visually very dominant. I think we can improve here to make them collapsable just like the [JSON schema](https://autodoc-pydantic.readthedocs.io/en/stable/users/configuration.html#show-schema-json).

I'm curious to get to know your thoughts on these points.

yves-renier commented:
- The watermark made the test harder as `erdantic` version is mentioned in it which we don't know in advance. I ignored that part of the output in the tests I implemented. Having a customizable label would also make those kind of tests much cleaner. I totally agree with you that the best would be an option from `erdantic` as I would neither be comfortable with removing it from `autodoc_pydantic`.
- about SVG format for the diagram, I just tested and it seems to works. I modified docs/source/conf.py accordingly.
- about making the diagram collapsible, I have no problem about that, I will make it the default and add an other option `model-erdantic-figure-collapsed` to disable it.

mansenfranzen commented on sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:
```diff
@@ -324,10 +333,48 @@ def add_collapsable_schema(self):
+            self.add_line('   ' + error_msg)
+            return
+
+        PydanticClass = self._get_pydantic_class()
```
Sphinx' autodoc-directives already hold a reference for the object to be documented via `self.object`. Please use `self.object` and drop `_get_pydantic_class()` completely.

mansenfranzen commented on sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:
```diff
@@ -324,10 +333,48 @@ def add_collapsable_schema(self):
+
+        """
+        source_name = self.get_sourcename()
+        if erd is None:
+            logger = sphinx.util.logging.getLogger(__name__)
+            error_msg = f'erdantic is not installed, no Entity Relationship Diagram for {self.fullname} is skipped.' + \
+                        ' See https://autodoc-pydantic.readthedocs.io/en/stable/users/installation.html'
+            logger.warning(error_msg, location="autodoc_pydantic")
+            self.add_line("..warning::")
+            self.add_line('   ' + error_msg)
+            return
```
Great usage of sphinx' logging and also adding a warning to the docs. However, I think it would be better to raise an exception here because if someone explicitly activates the ERD functionality, an error should be raised if it's not functioning. What do you think?

yves-renier replied:
That's was my initial thought, but I was not sure you would be happy with exceptions being raised. As it does not seem to be a problem, I will do it that way instead.

mansenfranzen commented on sphinxcontrib/autodoc_pydantic/directives/templates.py:
```diff
@@ -5,12 +5,10 @@
+   <p><details  class="{details_class}">
+   <summary>{summary}</summary>
 
-.. code-block:: json
-
-{}
+{lines}
```
Thumbs up for reusing the template :-)

mansenfranzen commented on pyproject.toml:
`erdantic= { version ="^0.5", optional = true }` needs to be added to `[tool.poetry.dependencies]`. Otherwise, the install with extras `erd` does not work (or at least did not work for me).

mansenfranzen commented on pyproject.toml:
```diff
@@ -52,6 +52,8 @@ dev = ["pytest",
        "flake8",
        "tox"]
```
Let's add `erdantic` to the `dev`, `test` and `docs` extras because it will always be required from now. This will also simplify `tox.ini`.

mansenfranzen commented on tox.ini:
`erd` extras can be removed if `erdantic` is added to all extras in `pyproject.toml`.

mansenfranzen commented on docs/source/users/configuration.rst:
```diff
@@ -195,6 +195,28 @@ Contains all modifications for pydantic `BaseModel`.
+   :values: True, False
+
+   Show the entity relationship diagram collapsed or not.
+   :ref:`model-erdantic-figure <autodoc_pydantic_model_erdantic_figure_collapsed>` must be True for this to have any effect.
```
The reference is to the configuration section itself (collapsed). It should be `autodoc_pydantic_model_erdantic_figure` instead.

mansenfranzen commented on .github/workflows/tests.yml:
The tests are not running currently because the `runs-on` is outdated with `runs-on: ubuntu-18.04`. Let's change this to `ubuntu-22.04` and see if the runners pick up the tests again.

mansenfranzen reviewed (changes requested):
As I said before, you've submitted great PR! I really appreciate the work you've put into the project.

I've added multiple small comments/recommendations. I hope they are self-explanatory. If not, please reach out.

I think we still need to include a short note/hint in the documentation about installing `erdantic` if the feature is enabled.

mansenfranzen commented:
@yves-renier By the way, great work adding `model-erdantic-figure-collapsed` along with tests!

yves-renier commented:
all tests passed on my repo (see https://github.com/yves-renier/autodoc_pydantic/actions/runs/4978871857/jobs/8909737895 ) except for the uploading of the documentation as I did not configure the upload token.

mansenfranzen commented:
@yves-renier Great work! I'm sorry for letting you wait so long. I was unfortunately busy otherwise.

Great job on adding the new sphinx versions to the test matrix while also fixing the corresponding sphinx compatibility issues. In fact, your PR is already worth multiple PRs because you've also fixed and updated the CI pipeline, too. 

There only minor points that I would like to comment on. I can happily finish it from here if you would allow me to commit on your fork ([see here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)). Otherwise, feel free to finish it yourself ;-).

After this PR is merged, I would be very curious to understand how you approached this PR. More specifically, did you use the developer's documentation? If so, was it helpful or did you miss something? In addition, you are by far the most involved committer towards the code base. Your feedback regarding design, structure and complexity would be highly appreciated. For example, what parts of the code base where not intuitive? What make you frown?

mansenfranzen commented on docs/source/users/configuration.rst:
```diff
@@ -195,6 +195,28 @@ Contains all modifications for pydantic `BaseModel`.
    Define the signature prefix for pydantic models.
 
 
+.. config_description:: autopydantic_model
+   :title: Show Erdantic figure
+   :path: target.configuration.ModelErdanticFigure
+   :confpy: autodoc_pydantic_model_erdantic_figure
+   :directive_option: model-erdantic-figure
+   :values: True, False
+
+   Show the entity relationship diagram of the schemas using erdantic.
+
+
+.. config_description:: autopydantic_model
+   :title: Show Erdantic figure collapsed
+   :path: target.configuration.ModelErdanticFigure
+   :confpy: autodoc_pydantic_model_erdantic_figure_collapsed
+   :directive_option: model-erdantic-figure-collapsed
+   :enable: model-erdantic-figure
+   :values: True, False
+
+   Show the entity relationship diagram collapsed or not.
+   :ref:`model-erdantic-figure <autodoc_pydantic_model_erdantic_figure>` must be True for this to have any effect.
+
```
I feel we should also include a short warning here about the `graphviz` dependency. You've added one in the user's installation section, but most will probably won't read it anyway and rather come here to understand how to enable the feature.

yves-renier replied:
implemented in 803e7c3

mansenfranzen commented on pyproject.toml:
Let's drop support for python 3.6 altogether. I think we can safely use 

```
[tool.poetry.dependencies]
python = ">=3.7.1,<4.0.0"
```

yves-renier replied:
fixed by 3e70a68

mansenfranzen commented on sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:
```diff
@@ -275,13 +279,19 @@ def add_content(self,
 
+        if self.pydantic.options.is_true("erdantic-figure", True):
+            if self.pydantic.options.is_true(
+                    "erdantic-figure-collapsed", True):
```
In my opinion, we can move `self.pydantic.options.is_true("erdantic-figure-collapsed", True)` into the `self.add_erdantic_figure`. This would remove the `if` condition here and also would make the `collapsed` argument of `self.add_erdantic_figure` obsolete. Overall, that would be a minor reduction of complexity.

yves-renier replied:
fixed by ff24325

mansenfranzen commented on sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:
```diff
@@ -324,10 +334,43 @@ def add_collapsable_schema(self):
+        lines.append("")
+
+        if collapsed:
+            lines = "\n".join(lines)
+            lines = TPL_COLLAPSE.format(
+                lines=lines,
+                summary="Show Entity Relationship Diagram",
+                details_class="autodoc_pydantic_collapsable_erd"
+            ).split("\n")
```
This block appears twice since we have the second feature that uses collapsible blocks (`autodoc_pydantic_collapsable_json`). To prevent code duplication, I suggest we use a function like this located in `sphinxcontrib/autodoc_pydantic/directives/templates.py`:
```python
def to_collapsable(lines: list[str], title, css_class) -> list[str]:
    """Place given lines into a collapsable HTML block.
    
    Parameters
    ----------
    lines: list
        The actual content, that should be placed into a collapsable block.
    title: str
        The name of the collapsable block title.
    css_class: str
        Name of the css class for the collapsable block.
        
    """
    
    lines = "\n".join(lines)
    lines = TPL_COLLAPSE.format(
        lines=lines,
        summary=title,
        details_class=css_class
    )
    return lines.split("\n")
```

Then, we could change the marked code to:

```python
if collapsed:
  lines = to_collapsable(
      lines=lines,
      title="Show Entity Relationship Diagram",
      css_class="autodoc_pydantic_collapsable_erd"
  )
```

yves-renier replied:
fixed in 67475f0

mansenfranzen commented on sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:
```diff
@@ -324,10 +334,43 @@ def add_collapsable_schema(self):
+            raise RuntimeError(error_msg)
+
+        # Graphviz [DOT language](https://graphviz.org/doc/info/lang.html)
+        figure_dot = erd.to_dot(self.object)
+        lines = [
+            ".. graphviz::",
+            "",
+        ]
+        lines += ['   ' + line
+                  for line in figure_dot.replace('\t', '   ').split('\n')]
+        lines.append("")
```
For my taste, I would like to make this part a bit more concise:

```python
figure_dot = erd.to_dot(self.object).replace('\t', '   ').split('\n')
lines_dot = ['   ' + line for line in figure_dot]
lines = [".. graphviz::", ""] + lines_dot + [""]
```

yves-renier replied:
done in 25af89a

mansenfranzen commented on tests/compatibility.py:
```diff
@@ -134,5 +136,15 @@ def module_doc_string_tab() -> str:
     return ""
 
+
+def get_type_expected(field_type: str):
+    if sphinx.version_info >= (6, 1):
+        optional_match = re.findall(r'Optional\[(\w*)\]', field_type)
+        if optional_match is not None and len(optional_match) > 0:
+            return optional_match[0] + ' | None'  # int | None
+    return field_type  # 'Optional[int]'
+
```
Could you please provide a simple doc string with an example of input -> output? This would greatly help to understand what this function does. Moreover, the function name itself is rather generic and requires some guessing.

yves-renier replied:
done in 1b7df9b

mansenfranzen reviewed (commented):
As I said before, excellent work! I've added some more comments but this PR would already qualify for merging.

In the end, we should not forget about the changelog. Surely I can take care of this if you want.

yves-renier commented:
Sorry, I totally missed the GH messages. Now I have seen it I will put back some effort into it.

>  if you would allow me to commit on your fork

I have done that during the PR creation, let me know if it does not work.

I will now go through your comments, then I will give you my feedback on the project.

yves-renier commented:
> In the end, we should not forget about the changelog. Surely I can take care of this if you want.

My last commit tried to address that. I tried to follow the format of the previous entries, but fell free to modify anything as you wish.

mansenfranzen commented:
> adding a link in the Documentation section of the Readme.md would not hurt.

Agreed - I will incorporate that in a separate release.

> One last thing, I did not manage to run sphinx in debug mode in my IDE (pycharm) which made the debugging of issues quite painful. If you know a way to achieve that, it could be a nice tip to add in the developer's documentation.

What was the error you were running into? Without pycharm's debugger I would be lost.

mansenfranzen commented:
I made several small modifications:

- Added a separate `extras` in pyproject.toml for `erdantic` as you've initially implemented it. It makes the user install and the CI setup with erdantic as a dependency both more explicit.
- Added a separate test running the test suite without installing erdantic since there were no tests without erdantic as a dependency anymore. Since we using tox, the package is build from scratch and installed in clean environment which helps to ensure that build/install process works fine.
- Updates docs/readme with versions/badges.

mansenfranzen commented:
Once again, thanks for all the great work you've put into this PR! It was a pleasure working with you :-).
